src/bulk_insert/bulk_insert.c

```diff
@@ -7,388 +7,211 @@
-            return NULL;
-        }
-        *argc -= attribute_count;
+    // Reusable buffer for storing properties of each node
+    SIValue values[prop_count];
+    unsigned int prop_num = 0;
 
-        label->attributes = malloc(sizeof(char*) * attribute_count);
-        for(int j = 0; j < attribute_count; j++) {
-            char *attribute = (char*)RedisModule_StringPtrLen(*argv++, NULL);
-            label->attributes[j] = attribute;   // Attribute is being duplicated by GraphEntity_Add_Properties.
+    while (data_idx < data_len) {
+        Node n;
+        Graph_CreateNode(gc->g, label_id, &n);
+        for (unsigned int i = 0; i < prop_count; i ++) {
+            values[i] = _BulkInsert_ReadProperty(data, &data_idx);
         }
+        GraphEntity_Add_Properties((GraphEntity*)&n, prop_count, prop_keys, values);
     }
 
-    return argv;
+    free(prop_keys);
+    return BULK_OK;
 }
 
-RedisModuleString** _BulkInsert_Parse_Labels(RedisModuleCtx *ctx, GraphContext *gc, LabelDesc *labels, int label_count,
-                                              RedisModuleString **argv, int *argc) {
-
-    // Consume labels.
-    for(int label_idx = 0; label_idx < label_count; label_idx++) {
-        argv = _BulkInsert_Parse_Label(ctx, &labels[label_idx], argv, argc);
-        if(argv == NULL) {
-            // Free previous parsed labels.
-            for(int i = 0; i < label_idx; i++) {
-                if(labels[i].attribute_count > 0) {
-                    free(labels[i].attributes);
-                }
-            }
-            return NULL;
-        }
+int _BulkInsert_ProcessRelationFile(RedisModuleCtx *ctx, GraphContext *gc, const char *data, size_t data_len) {
+    size_t data_idx = 0;
 
-        LabelStore *store = GraphContext_GetStore(gc, labels[label_idx].label, STORE_NODE);
-        LabelStore *allStore = GraphContext_AllStore(gc, STORE_NODE);
-        if(store == NULL) {
-            store = GraphContext_AddLabel(gc, labels[label_idx].label);
-        }
+    int reltype_id;
+    unsigned int prop_count;
+    // Read property keys from header and update schema
+    char **prop_keys = _BulkInsert_ReadHeader(gc, STORE_EDGE, data, &data_idx, &reltype_id, &prop_count);
 
-        labels[label_idx].label_id = store->id;
-        LabelStore_UpdateSchema(store, labels[label_idx].attribute_count, labels[label_idx].attributes);
-        LabelStore_UpdateSchema(allStore, labels[label_idx].attribute_count, labels[label_idx].attributes);
-    }
+    // Reusable buffer for storing properties of each relation 
+    SIValue values[prop_count];
+    unsigned int prop_num = 0;
+    NodeID src;
+    NodeID dest;
 
-    return argv;
-}
+    while (data_idx < data_len) {
+        Edge e;
+        // Next 8 bytes are source ID
+        src = *(NodeID*)&data[data_idx];
+        data_idx += sizeof(NodeID);
+        // Next 8 bytes are destination ID
+        dest = *(NodeID*)&data[data_idx];
+        data_idx += sizeof(NodeID);
```

Comment:
I think we need to make sure we do not read pass data_len, we assume here it is safe to increase data_idx

Reply:
The `while (data_idx < data_len)` condition can only be violated within the loop if the binary token abruptly ends in the middle of describing a single node or relation. The script should be incapable of making that error, and I think introducing branching logic at every data read would negatively impact our performance. Thoughts?